docs/middleware.md

```diff
@@ -395,12 +395,111 @@ class ASGIMiddleware:
 
     async def __call__(self, scope, receive, send):
         if scope["type"] == "http":
-            request = Request(scope, receive, send)
-            # Do something here!
+            request = Request(scope)
+            ... # Use `request.method`, `request.url`, `request.headers`, etc.
+
+        await self.app(scope, receive, send)
+```
+
+You can also reuse [responses](responses.md), which are ASGI applications as well.
+
+#### Sending eager responses
+
+Inspecting the connection `scope` allows you to conditionally call into a different ASGI app. One use case might be sending a response without calling into the app.
+
+As an example, this middleware uses a dict to perform permanent redirects based on the requested path. This could be used to implement ongoing support of legacy URLs in case you need to refactor route URL patterns.
+
+```python
+from starlette.datastructures import URL
+from starlette.responses import RedirectResponse
+
+class RedirectsMiddleware:
+    def __init__(self, app, path_mapping: dict):
+        self.app = app
+        self.path_mapping = path_mapping
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        url = URL(scope=scope)
+
+        if url.path in self.path_mapping:
+            url = url.replace(path=self.path_mapping[url.path])
+            response = RedirectResponse(url, status_code=301)
+            await response(scope, receive, send)
+            return
+
         await self.app(scope, receive, send)
 ```
 
-Or we might use `MutableHeaders` to change the response headers:
+Example usage would look like this:
+
+```python
+from starlette.applications import Starlette
+from starlette.middleware import Middleware
+
+routes = ...
+
+redirections = {
+    "/v1/resource/": "/v2/resource/",
+    # ...
+}
+
+middleware = [
+    Middleware(RedirectsMiddleware, path_mapping=redirections),
+]
+
+app = Starlette(routes=routes, middleware=middleware)
+```
+
+
+#### Inspecting or modifying the request
+
+Request information can be accessed or changed by manipulating the `scope`. For a full example of this pattern, see Uvicorn's [`ProxyHeadersMiddleware`](https://github.com/encode/uvicorn/blob/master/uvicorn/middleware/proxy_headers.py) which inspects and tweaks the `scope` when serving behind a frontend proxy.
```

Comment:
Not sure if relevant... But maybe using the permalink? https://github.com/encode/uvicorn/blob/b21ecabc5bf911f571e0629438315a1e5472065c/uvicorn/middleware/proxy_headers.py